### ISSUE TITLE
continual learning cycle - how to continually train models on "future" data?

### ISSUE THREAD

**Issue by matkowski-voy** [ContinualAI/avalanche#1070]
Hi Everyone,

First, thanks for the Avalanche, I think it is a great library to experiment with many CL methods.
This is rather general issue probably related to this one: #1016. 

In many real continual learning use cases, we don't have all data at the same time but rather than they are coming at different times in future. Thus, fixing a dataset, dataloaders, scenarios and CL strategy in the "beginning" is not possible. 

**Avalanche examples use case:** Most of Avalanche examples, use well predefined benchmarks from the research papers and fix the dataset, dataloaders, scenarios and CL strategy before executing the for loop. 
**My use case** Now, I have been trying to run one simple "imaginary" use case, where, I already have a pretrained model from the past, but most of data it has been trained on are not available thus I can only start continual learning strategy from the pretrained model, with limited past data and the new data that I have at hand now (and likely won't have in few months). 

I run two experiments using memory replay method (replay plugin):
**exp1** Avalanche examples use case -> train on experience A (old data) and then B (new data) using the for loop and predefined scenario to get the **model_cl1**. After training on first experience A, I get **model_A**. Two iterations to do CL here.
**exp2** My use case -> get **model_old** , which has been trained on experience A in the past and no train it on experience B only (new data) using continual learning strategy to get the **model_old_B**. One iteration to do CL here, because model_old is already available. 

The results are:
1) **model_old** and **model_A** are the same. It is expected because they are the same models trained on the same data.
2) **model_cl1** and **model_old_B** are different. It is not necessarily expected because 1) is true and CL strategy step on experience B is done on the same data. 
This would suggest that in 1), CL strategy already has access to "something", that is no yet there (because it will be in future) in the training stream.

Please let me know if this is the case and there is any workaround or I am simply wrong and not using Avalanche CL strategy correctly. 

Cheers and thanks,
Woj

**Comment by AntonioCarta** [comment]
I don't completely understand what you want but I think it's doable. In Avalanche you can:
- start a strategy from a pretrained model (e.g. `strat = Naive(model_a, ...); strat.train(new_stream)`.
- create a stream by combining different datasets, so if you need a non-standard benchmark you can make it.

**Comment by matkowski-voy** [comment]
Hi Antonio, thanks for the answer.
I tried to keep my question as general as possible, sorry if that makes some confusion. So here, to clarify in one particular example I attached the code with exp1 (avalanche use case training in a loop) and exp2 (my use case). 
`import torch
from torch.utils.data import TensorDataset
from avalanche.training.strategies import BaseStrategy
from avalanche.training.plugins import ReplayPlugin
from avalanche.evaluation.metrics import forgetting_metrics, accuracy_metrics,\
    loss_metrics, confusion_matrix_metrics
from avalanche.models import SimpleMLP
import numpy as np
from avalanche.logging import InteractiveLogger
from avalanche.training.plugins import EvaluationPlugin
from avalanche.benchmarks.generators import dataset_benchmark
                                            
import torch.nn as nn
import torch.optim as optim
import copy

torch.manual_seed(0) 
np.random.seed(0) 
torch.backends.cudnn.deterministic = True

device = torch.device('cpu')


N1 = 100
N2 = 100

x_data1 = torch.rand(N1,784)*2
y_data1 = torch.randint(0, 2, (N1,))

x_data2 = torch.rand(N2,784)
y_data2 = torch.randint(0, 2, (N2,))

trainset1 = TensorDataset(x_data1, y_data1)
trainset2 = TensorDataset(x_data2, y_data2)

x_data1 = torch.rand(N1,784)*2
y_data1 = torch.randint(0, 2, (N1,))

x_data2 = torch.rand(N2,784)
y_data2 = torch.randint(0, 2, (N2,))

testset1 = TensorDataset(x_data1, y_data1)
testset2 = TensorDataset(x_data2, y_data2)


generic_scenario = dataset_benchmark([trainset1, trainset2],
                                     [testset1, testset2])

eval_plugin = EvaluationPlugin(
    accuracy_metrics(minibatch=True, epoch=True, experience=True, stream=True),
    loss_metrics(minibatch=True, epoch=True, experience=True, stream=True),
    forgetting_metrics(experience=True, stream=True),
    confusion_matrix_metrics(num_classes=2, save_image=False, stream=True),
    loggers=[InteractiveLogger()],
    strict_checks=False
)


print('Avalanche use-case example (exp1 example)')
net = SimpleMLP(2)
optimizer = optim.SGD(net.parameters(), lr=0.01, momentum=0.9)
criterion = nn.CrossEntropyLoss()
replay = ReplayPlugin(mem_size=10)
cl_strategy = BaseStrategy(
    net, optimizer, criterion,
    plugins=[replay],device=device,train_mb_size=2,train_epochs=1,evaluator=eval_plugin)
results = []
for i, experience in enumerate(generic_scenario.train_stream,0):

    print("Start of experience: ", experience.current_experience)

    cl_strategy.train(experience)
    print('Training completed')

    print('Computing accuracy on the whole test set')
    results.append(cl_strategy.eval(generic_scenario.test_stream))


print('-')
print('-')
print('----------------------------------------------------------------------')
print('----------------------------------------------------------------------')
print('-')
print('-')
print('my use-case example (exp2 example)')
torch.manual_seed(0) 
np.random.seed(0) 
torch.backends.cudnn.deterministic = True
results = []   
net = SimpleMLP(2)
optimizer = optim.SGD(net.parameters(), lr=0.01, momentum=0.9)
criterion = nn.CrossEntropyLoss() 
cl_strategy = BaseStrategy(
    net, optimizer, criterion,device=device,train_mb_size=2,train_epochs=1,evaluator=eval_plugin)

cl_strategy.train(generic_scenario.train_stream[0])
print('Computing accuracy on the whole test set')
results.append(cl_strategy.eval(generic_scenario.test_stream))
replay = ReplayPlugin(mem_size=10)
#net = copy.deepcopy(cl_strategy.model)
#optimizer = copy.deepcopy(cl_strategy.optimizer)
cl_strategy = BaseStrategy(
    net, optimizer, criterion,
    plugins=[replay],device=device,train_mb_size=2,train_epochs=1,evaluator=eval_plugin)
cl_strategy.train(generic_scenario.train_stream[1]) 
print('Computing accuracy on the whole test set')
results.append(cl_strategy.eval(generic_scenario.test_stream))`

What I would expect?
1. Same metrics after training experience 0 -> result is as expected.
2. Same metrics after training on experience 1 -> result is not as expected
Log:
Avalanche use-case example
Start of experience:  0
-- >> Start of training phase << --
-- Starting training on experience 0 (Task 0) from train stream --
 94%|████████████████████████████████████████████████████████████████████████████▏    | 47/50 [00:00<00:00, 274.54it/s]C:\Users\t0249889\Anaconda3\envs\opencv\lib\site-packages\avalanche\training\plugins\evaluation.py:84: UserWarning: No benchmark provided to the evaluation plugin. Metrics may be computed on inconsistent portion of streams, use at your own risk.
  warnings.warn(
100%|█████████████████████████████████████████████████████████████████████████████████| 50/50 [00:00<00:00, 276.25it/s]
Epoch 0 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 5.0206
	Loss_MB/train_phase/train_stream/Task000 = 0.7647
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.5400
	Top1_Acc_MB/train_phase/train_stream/Task000 = 0.5000
-- >> End of training phase << --
Training completed
Computing accuracy on the whole test set
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 568.18it/s]
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.7736
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4400
-- Starting eval on experience 1 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 561.79it/s]
> Eval on experience 1 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 0.6884
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.5500
-- >> End of eval phase << --
	ConfusionMatrix_Stream/eval_phase/test_stream = 
tensor([[ 99,   0],
        [101,   0]])
	Loss_Stream/eval_phase/test_stream/Task000 = 0.7310
	StreamForgetting/eval_phase/test_stream = 0.0000
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.4950
Start of experience:  1
-- >> Start of training phase << --
-- Starting training on experience 1 (Task 0) from train stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 271.74it/s]
Epoch 0 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.7416
	Loss_MB/train_phase/train_stream/Task000 = 0.6964
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.4950
	Top1_Acc_MB/train_phase/train_stream/Task000 = 0.5000
-- >> End of training phase << --
Training completed
Computing accuracy on the whole test set
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 558.67it/s]
> Eval on experience 0 (Task 0) from test stream ended.
	ExperienceForgetting/eval_phase/test_stream/Task000/Exp000 = -0.1200
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.6870
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.5600
-- Starting eval on experience 1 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 564.96it/s]
> Eval on experience 1 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 0.7030
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.4500
-- >> End of eval phase << --
	ConfusionMatrix_Stream/eval_phase/test_stream = 
tensor([[  0,  99],
        [  0, 101]])
	Loss_Stream/eval_phase/test_stream/Task000 = 0.6950
	StreamForgetting/eval_phase/test_stream = -0.1200
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.5050
-
-
----------------------------------------------------------------------
----------------------------------------------------------------------
-
-
my use-case example
-- >> Start of training phase << --
-- Starting training on experience 0 (Task 0) from train stream --
100%|█████████████████████████████████████████████████████████████████████████████████| 50/50 [00:00<00:00, 295.86it/s]
Epoch 0 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 59.7577
	Loss_MB/train_phase/train_stream/Task000 = 0.8599
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.4600
	Top1_Acc_MB/train_phase/train_stream/Task000 = 0.0000
-- >> End of training phase << --
Computing accuracy on the whole test set
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 568.17it/s]
> Eval on experience 0 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.6833
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.5800
-- Starting eval on experience 1 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 555.55it/s]
> Eval on experience 1 (Task 0) from test stream ended.
	ExperienceForgetting/eval_phase/test_stream/Task000/Exp001 = -0.0300
	Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 0.6945
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.4800
-- >> End of eval phase << --
	ConfusionMatrix_Stream/eval_phase/test_stream = 
tensor([[35, 64],
        [30, 71]])
	Loss_Stream/eval_phase/test_stream/Task000 = 0.6889
	StreamForgetting/eval_phase/test_stream = -0.0300
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.5300
-- >> Start of training phase << --
-- Starting training on experience 1 (Task 0) from train stream --
100%|█████████████████████████████████████████████████████████████████████████████████| 50/50 [00:00<00:00, 289.02it/s]
Epoch 0 ended.
	Loss_Epoch/train_phase/train_stream/Task000 = 0.7162
	Loss_MB/train_phase/train_stream/Task000 = 0.7445
	Top1_Acc_Epoch/train_phase/train_stream/Task000 = 0.4500
	Top1_Acc_MB/train_phase/train_stream/Task000 = 0.0000
-- >> End of training phase << --
Computing accuracy on the whole test set
-- >> Start of eval phase << --
-- Starting eval on experience 0 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 564.96it/s]
> Eval on experience 0 (Task 0) from test stream ended.
	ExperienceForgetting/eval_phase/test_stream/Task000/Exp000 = 0.1400
	Loss_Exp/eval_phase/test_stream/Task000/Exp000 = 0.6994
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp000 = 0.4400
-- Starting eval on experience 1 (Task 0) from test stream --
100%|███████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 568.19it/s]
> Eval on experience 1 (Task 0) from test stream ended.
	Loss_Exp/eval_phase/test_stream/Task000/Exp001 = 0.6899
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp001 = 0.5500
-- >> End of eval phase << --
	ConfusionMatrix_Stream/eval_phase/test_stream = 
tensor([[ 99,   0],
        [101,   0]])
	Loss_Stream/eval_phase/test_stream/Task000 = 0.6947
	StreamForgetting/eval_phase/test_stream = 0.1400
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.4950

**Comment by AntonioCarta** [comment]
exp2 is wrong. In the second `train` call you use a `ReplayPlugin`, but the buffer is empty because it has never seen the first experience.

**Comment by matkowski-voy** [comment]
Hi Antonio, 

Thanks for answering :)
Please take a look 

A: "In the second train call you use a ReplayPlugin, but the buffer is empty because it has never seen the first experience"  
W:  then, how can I create and pass the buffer from the first experience to the second one without defining the training strategy at the beginning and running the for loop?

A: "exp2 is wrong." 
W: I am not sure what you mean by wrong, but for example, in the code above, you could replace ReplayPlugin with EWC and the outcome is similar, which is the same accuracy after experience 0 and then different after experience 1.

So, we go back to my original problem, that I mentioned above in my first post.
In many real continual learning use cases, we don't have all data at the same time but rather than they are coming at different times in future. Thus, fixing a dataset, dataloaders, scenarios and CL strategy in the "beginning" is not possible.

This means that we cannot create a data stream once (because future data are not available yet) and train model in a for loop on this stream, but:
- train model_A on data A -> save model_A (and perhaps something more things necessary to use in future CL)
- load model_A and continue to train it using some selected CL method on new data B (perhaps without data A available) -> save model_B
- load model_B and continue to train it using some selected CL method  on new data C (perhaps without data A and B available) -> save model_C
- etc. 

There is never access to the full data stream A,B,C but they appear and disappear in a time sequence. If there is access to full data stream A,B,C at the same time, then one would use joint training because would likely provide the highest accuracy.   

Is it possible to use Avalanche for such use case where there is no access to the full data stream and one time and we need to save models after training and then load them for the new training at different times?

Thanks :)

**Comment by AntonioCarta** [comment]
- you can use lazy streams/iterators, so you case (1) is somewhat supported.
- what is not supported yet is serialization, so you cannot stop a strategy, save it in memory, and restart at a later time. We are working on adding support for serialization.
